code/crates/core-consensus/src/handle/vote_set.rs

```diff
@@ -33,18 +37,31 @@ pub async fn on_vote_set_response<Ctx>(
     co: &Co<Ctx>,
     state: &mut State<Ctx>,
     metrics: &Metrics,
-    response: VoteSet<Ctx>,
+    vote_set: VoteSet<Ctx>,
+    polka_certificate: Option<PolkaCertificate<Ctx>>,
 ) -> Result<(), Error<Ctx>>
 where
     Ctx: Context,
 {
     debug!(
-        height = %state.height(), round = %state.round(), votes.count = %response.len(),
+        height = %state.height(), round = %state.round(),
+        votes.count = %vote_set.len(),
+        polka.certificate = %polka_certificate.is_some(),
         "Received vote set response"
     );
 
-    for vote in response.votes {
-        let _ = on_vote(co, state, metrics, vote).await;
+    if let Some(certificate) = polka_certificate {
+        apply_driver_input(
+            co,
+            state,
+            metrics,
+            DriverInput::PolkaCertificate(certificate),
+        )
+        .await?;
+    } else {
+        for vote in vote_set.votes {
```

Comment:
@ancazamfir I don't believe we need to process both the votes and the certificate, but perhaps I am mistaken?

Reply:
Do you mean it's fine to process both?